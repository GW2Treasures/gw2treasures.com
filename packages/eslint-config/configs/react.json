{
  "$schema": "https://json.schemastore.org/eslintrc",

  "extends": [
    "plugin:react/recommended",
    "@gw2treasures/eslint-config"
  ],

  "rules": {
    // `<C foo='bar'>` → `<C foo="bar">`
    "jsx-quotes": ["warn", "prefer-double"],

    // `< C / >` → `<C/>`
    "react/jsx-tag-spacing": ["warn", {
      "closingSlash": "never",
      "beforeSelfClosing": "never",
      "afterOpening": "never",
      "beforeClosing": "never"
    }],

    // `<C prop={'test'}/>` → `<C prop="test"/>`
    "react/jsx-curly-brace-presence": ["warn", "never"],

    // add parens around jsx
    "react/jsx-wrap-multilines": ["warn", {
      "declaration": "parens-new-line",
      "assignment": "parens-new-line",
      "return": "parens-new-line",
      "arrow": "parens-new-line",
      "condition": "parens-new-line",
      "logical": "parens-new-line",
      "prop": "parens-new-line"
    }],

    // multiline closing bracket location
    "react/jsx-closing-bracket-location": ["warn", {
      "nonEmpty": "line-aligned",
      "selfClosing": "after-props"
    }],

    // indent jsx with 2 spaces
    "react/jsx-indent": ["warn", 2, { "checkAttributes": true, "indentLogicalExpressions": true }],

    // indent props with 2 spaces
    "react/jsx-indent-props": ["warn", 2],

    // disallows using the array index as key
    "react/no-array-index-key": "warn",

    // `<Foo></Foo>` → `<Foo/>`
    "react/self-closing-comp": "warn",

    // `<Foo bar={true}>` → `<Foo bar/>`
    "react/jsx-boolean-value": "warn",

    // align the closing JSX tag with the opening tag
    "react/jsx-closing-tag-location": "warn",

    // `<React.Fragment>` → `<>`
    "react/jsx-fragments": "warn",

    // require key
    "react/jsx-key": ["warn", { "checkFragmentShorthand": true }],

    // prevent multiple spaces in jsx
    "react/jsx-props-no-multi-spaces": "warn",

    // `<Foo bar={ baz }>` → `<Foo bar={baz}/>`
    "react/jsx-curly-spacing": "warn",

    // require dependencies to be in package.json
    "import/no-extraneous-dependencies": "error",

    // disable next rules
    "@next/next/no-html-link-for-pages": "off"
  },

  "overrides": [
    {
      "files": ["**/*.ts", "**/*.tsx"],
      "rules": {
        "@typescript-eslint/type-annotation-spacing": "warn",

        "@typescript-eslint/naming-convention": ["warn", {
          "selector": "default", "format": null
        },{
          "selector": "typeLike", "format": ["PascalCase"]
        }]
      }
    }
  ]
}
