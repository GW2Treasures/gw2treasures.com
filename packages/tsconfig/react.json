{
  "$schema": "https://json.schemastore.org/tsconfig",
  "extends": "./base.json",
  "compilerOptions": {
    "lib": ["dom", "dom.iterable", "esnext"],
    "allowJs": true,
    "jsx": "preserve",
    "importHelpers": true,
    "incremental": true,

    "noEmit": true,
    "declaration": false
  }
}
