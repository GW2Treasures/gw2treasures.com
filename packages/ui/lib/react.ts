import type { Ref } from 'react';

export interface RefProp<T = Element> {
  ref?: Ref<T>
}
