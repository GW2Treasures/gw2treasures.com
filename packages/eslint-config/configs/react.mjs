import ts from 'typescript-eslint';
import react from 'eslint-plugin-react';
import config from './index.mjs';

export default ts.config(
  react.configs.flat.recommended,
  react.configs.flat['jsx-runtime'],
  config,
  {
    rules: {
      // `<C foo='bar'>` → `<C foo="bar">`
      "@stylistic/jsx-quotes": ["warn", "prefer-double"],
        
      // `< C / >` → `<C/>`
      "@stylistic/jsx-tag-spacing": ["warn", {
        "closingSlash": "never",
        "beforeSelfClosing": "never",
        "afterOpening": "never",
        "beforeClosing": "never"
      }],

      // `<C prop={'test'}/>` → `<C prop="test"/>`
      "@stylistic/jsx-curly-brace-presence": ["warn", "never"],

      // add parens around jsx
      "@stylistic/jsx-wrap-multilines": ["warn", {
        "declaration": "parens-new-line",
        "assignment": "parens-new-line",
        "return": "parens-new-line",
        "arrow": "parens-new-line",
        "condition": "parens-new-line",
        "logical": "parens-new-line",
        "prop": "parens-new-line"
      }],

      // multiline closing bracket location
      "@stylistic/jsx-closing-bracket-location": ["warn", {
        "nonEmpty": "line-aligned",
        "selfClosing": "after-props"
      }],

      // indent jsx with 2 spaces
      "@stylistic/jsx-indent": ["warn", 2, { "checkAttributes": true, "indentLogicalExpressions": true }],

      // indent props with 2 spaces
      "@stylistic/jsx-indent-props": ["warn", 2],

      // disallows using the array index as key
      "react/no-array-index-key": "warn",

      // `<Foo></Foo>` → `<Foo/>`
      "react/self-closing-comp": "warn",

      // `<Foo bar={true}>` → `<Foo bar/>`
      "react/jsx-boolean-value": "warn",

      // align the closing JSX tag with the opening tag
      "@stylistic/jsx-closing-tag-location": "warn",

      // `<React.Fragment>` → `<>`
      "react/jsx-fragments": "warn",

      // require key
      "react/jsx-key": ["warn", { "checkFragmentShorthand": true }],

      // prevent multiple spaces in jsx
      "@stylistic/jsx-props-no-multi-spaces": "warn",

      // `<Foo bar={ baz }>` → `<Foo bar={baz}/>`
      "@stylistic/jsx-curly-spacing": "warn",
    }
  }
);
