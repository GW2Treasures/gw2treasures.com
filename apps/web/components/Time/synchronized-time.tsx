'use client';

import { useInterval } from '@/lib/useInterval';
import { createContext, startTransition, use, useState, type FC, type ReactNode } from 'react';

// this context is used to used to have a single interval updating the time
// otherwise it could happen that multiple timers update slightly offset
export const SynchronizedTimeContext = createContext<Date | undefined>(undefined);

export interface SynchronizedTimeProviderProps {
  children: ReactNode,
}

export const SynchronizedTimeProvider: FC<SynchronizedTimeProviderProps> = ({ children }) => {
  // the time is undefined during SSR to avoid hydration warnings and allow caching
  const [time, setTime] = useState<Date>();

  // TODO: stop interval if page is not visible (`useVisibilityState`)
  useInterval(() => {
    startTransition(() => setTime(new Date()));
  }, 1000);

  return (
    <SynchronizedTimeContext value={time}>
      {children}
    </SynchronizedTimeContext>
  );
};

export function useSynchronizedTime() {
  return use(SynchronizedTimeContext);
}
