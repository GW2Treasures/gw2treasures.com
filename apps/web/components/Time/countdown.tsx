'use client';

import type { FC, ReactNode } from 'react';
import { useSynchronizedTime } from './synchronized-time';
import { Skeleton } from '../Skeleton/Skeleton';

export interface CountDownProps {
  /** UTC offset of repeating event */
  offsetMinutes: number,

  /** How many minutes between repeats */
  repeatMinutes: number,

  /** Highlight time (bold) if next event is within `highlightNextMinutes` minutes */
  highlightNextMinutes?: number,

  /** How many minutes the event stays active for and renders `active` instead of a countdown */
  activeDurationMinutes: number,

  /** Node to render when the event is active */
  active: ReactNode,
}

export const CountDown: FC<CountDownProps> = ({ offsetMinutes, repeatMinutes, highlightNextMinutes, activeDurationMinutes, active }) => {
  const time = useSynchronizedTime();

  // don't render the timer on the server
  if(!time) {
    return <Skeleton width={64}/>;
  }

  // get minutes since midnight UTC
  const currentHours = time.getUTCHours();
  const currentMinutes = currentHours * 60 + time.getUTCMinutes();

  // calculate how many minutes have passed since the last event
  const minutesSinceLast = (currentMinutes - offsetMinutes) % repeatMinutes;

  // render active component if event is active
  const isActive = minutesSinceLast < activeDurationMinutes;
  if(isActive) {
    return active;
  }

  // calculate next event
  const minutesUntilNext = repeatMinutes - minutesSinceLast;

  const remainingHours = Math.floor((minutesUntilNext - 1) / 60);
  const remainingMinutes = (60 + minutesUntilNext - 1) % 60;
  const remainingSeconds = (60 - time.getSeconds() - 1) % 60;

  // highlight countdown
  const Wrap = highlightNextMinutes && remainingMinutes < highlightNextMinutes ? 'strong' : 'span';

  return (
    <Wrap>
      in {remainingHours > 0 ? remainingHours + ':' : ''}{zeroPad(remainingMinutes)}:{zeroPad(remainingSeconds)}
    </Wrap>
  );
};

function zeroPad(number: number) {
  return number.toString().padStart(2, '0');
}
