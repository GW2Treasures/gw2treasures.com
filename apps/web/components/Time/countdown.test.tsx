import { render } from '@testing-library/react';
import { CountDown, type CountDownProps } from './countdown';
import { SynchronizedTimeContext } from './synchronized-time';

const defaultProps: CountDownProps = {
  active: 'active',
  activeDurationMinutes: 1,
  offsetMinutes: 0,
  repeatMinutes: 60,
  highlightNextMinutes: 1
};

const testCases: ({ time: Date | undefined } & Partial<CountDownProps>)[] = [
  { time: undefined },
  { time: new Date('2025-09-16T00:00:00.000Z') },
  { time: new Date('2025-09-16T00:00:59.999Z') },
  { time: new Date('2025-09-16T00:01:00.000Z') },
  { time: new Date('2025-09-16T00:59:00.000Z') },
  { time: new Date('2025-09-16T00:59:59.999Z') },
  { time: new Date('2025-09-16T01:00:00.000Z') },
  { time: new Date('2025-09-16T01:00:00.000Z'), repeatMinutes: 120 },
  { time: new Date('2025-09-16T00:30:00.000Z'), repeatMinutes: 120 },
];

describe('CountDown', () => {
  it.each(testCases)('test', ({ time, ...props }) => {
    expect(
      render(
        <SynchronizedTimeContext value={time}>
          <CountDown {...defaultProps} {...props}/>,
        </SynchronizedTimeContext>
      ).asFragment()
    ).toMatchSnapshot();
  });
});
