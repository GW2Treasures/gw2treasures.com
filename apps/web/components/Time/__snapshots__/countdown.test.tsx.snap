// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CountDown test 1`] = `
<DocumentFragment>
  <span
    class="skeleton"
    style="width: 64px;"
  />
  ,
</DocumentFragment>
`;

exports[`CountDown test 2`] = `
<DocumentFragment>
  active,
</DocumentFragment>
`;

exports[`CountDown test 3`] = `
<DocumentFragment>
  active,
</DocumentFragment>
`;

exports[`CountDown test 4`] = `
<DocumentFragment>
  <span>
    in 58:59
  </span>
  ,
</DocumentFragment>
`;

exports[`CountDown test 5`] = `
<DocumentFragment>
  <strong>
    in 00:59
  </strong>
  ,
</DocumentFragment>
`;

exports[`CountDown test 6`] = `
<DocumentFragment>
  <strong>
    in 00:00
  </strong>
  ,
</DocumentFragment>
`;

exports[`CountDown test 7`] = `
<DocumentFragment>
  active,
</DocumentFragment>
`;

exports[`CountDown test 8`] = `
<DocumentFragment>
  <span>
    in 59:59
  </span>
  ,
</DocumentFragment>
`;

exports[`CountDown test 9`] = `
<DocumentFragment>
  <span>
    in 1:29:59
  </span>
  ,
</DocumentFragment>
`;
